### PR TITLE
Feature/thread date separator

### PR DESCRIPTION
**Pull request by zacheryconverse**
# Submit a pull request

### 🎯 Goal

Create the ability to add date separator components into the `Thread`

### 🛠 Implementation details

pass enableThreadDateSeparator to `Thread` which passes the prop to `MessageList` or `VirtualizedMessageList`

### 🎨 UI Changes

Screen Shot 2022-03-15 at 4 53 43 PM

### CODE REVIEW

**Review by petyosi — APPROVED** [review]

**Comment by MartinCupela** [comment]
@zacheryconverse @petyosi  there were added some conditions in `processMessage`. Shouldn't we add tests?

File: src/components/MessageList/utils.ts [hunk 1]
```diff
@@ -71,18 +73,23 @@ export const processMessages = <
     if (
       i > 0 &&
       !disableDateSeparator &&
-      !threadList &&
       previousMessage.created_at &&
-      isDate(previousMessage.created_at)
+      isDate(previousMessage.created_at) &&
+      (!threadList || (threadList && enableThreadDateSeparator))
```
Comment (arnautov-anton) [inline src/components/MessageList/utils.ts:78]: These conditions get cryptic and hard to understand at times - I wouldn't mind in-code eplanation comments - what do you guys think? @MartinCupela @petyosi

Comment (petyosi) [reply src/components/MessageList/utils.ts:78]: Agreed. @arnautov-anton feel ree to add a suggestion. I usually create extra variables with descriptive names as means to document intent.

Comment (MartinCupela) [reply src/components/MessageList/utils.ts]: @arnautov-anton @petyosi I have added comments. Not just to explain the conditions, but also to document the `processMessages` function and its params + tests.

**Review by MartinCupela — APPROVED** [review]

File: src/components/MessageList/utils.ts [hunk 2]
```diff
@@ -10,31 +10,52 @@ import type { StreamMessage } from '../../context/ChannelStateContext';
...
   StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
 > = {
-  disableDateSeparator: boolean;
-  hideDeletedMessages: boolean;
-  hideNewMessageSeparator: boolean;
   messages: StreamMessage<StreamChatGenerics>[];
   userId: string;
+  /** Disable date separator in main message list */
+  disableDateSeparator?: boolean;
+  /** Enable date separator in main message list. Has to be accompanied by @param threadList. */
```
Comment (zacheryconverse) [inline src/components/MessageList/utils.ts:17]: /** Enable date separator in ~~main~~ THREAD message list. Has to be accompanied by @param threadList. */
I believe this should be changed from 'main message list' to 'thread message list'

File: src/components/MessageList/utils.ts [hunk 3]
```diff
@@ -11,6 +11,7 @@ type ProcessMessagesParams<
   StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics
 > = {
   disableDateSeparator: boolean;
+  enableThreadDateSeparator: boolean;
   hideDeletedMessages: boolean;
   hideNewMessageSeparator: boolean;
```
[the remark this answers is not in the document]
Comment (arnautov-anton) [reply src/components/MessageList/utils.ts]: The more I look at it the less I understand. Do we really need properties `threadList` and `enableThreadDateSeparator` in the `processMessages` function? Shouldn't be the function universal to any list? What I mean is that both `Channel` and `Thread` render either `VirtualizedMessageList` or `MessageList` which means both of these lists use identical messages (and their properties). The function should only accept `enableDateSeparators` (or `disableDateSeparators`, whichever property was introduced first as to not break stuff) and aforementioned components should do the deciding meaning `Channel` or `MessageList` rendered within would accept property `enableDateSeparators` which would be set to `true` by default in `MessageList` and `VirtualizedMessageList` and modified by `Thread` component which would set this property to `false` by default.

```tsx
<Channel>
	<Window>
		<MessageList /> // enableDateSeparator is set to `true` by default
	</Window>
	// here it has to be set explicitly as it defaults to `false`
	// this value is being passed down to whatever message list component the Thread component is rendering
	// meaning either VML or ML
	<Thread enableDateSeparator={true} /> 
</Channel>
```

@MartinCupela @petyosi thoughts?

[the remark this answers is not in the document]
Comment (MartinCupela) [reply src/components/MessageList/utils.ts]: @arnautov-anton Yes, it should be that way. I will take another look.

[the remark this answers is not in the document]
Comment (MartinCupela) [reply src/components/MessageList/utils.ts]: on it

[the remark this answers is not in the document]
Comment (petyosi) [reply src/components/MessageList/utils.ts]: I agree with Anton here, as far as . Ideally, noth message lists should not "know" if they are in threaded mode, and in extension, the processMessages itself. 

Also, if possible, the public flags should be consistent in their naming (both should be `enable*`), but have different default values.

**Review by arnautov-anton — APPROVED** [review]
LGTM :taco:

**Pull request merged**